File: smash-ranks-android/app/src/main/java/com/garpr/android/networking/ApiCall.java
package com.garpr.android.networking;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.lang.ref.WeakReference;

public class ApiCall<T> implements ApiListener<T> {

    private final WeakReference<ApiListener<T>> mReference;


    public ApiCall(@NonNull final ApiListener<T> listener) {
        mReference = new WeakReference<>(listener);
    }

    @Override
    public void failure() {
        final ApiListener<T> listener = mReference.get();

        if (listener != null && listener.isAlive()) {
            listener.failure();
        }
    }

    @Override
    public boolean isAlive() {
        final ApiListener<T> listener = mReference.get();
        return listener != null && listener.isAlive();
    }

    @Override
    public void success(@Nullable final T object) {
        final ApiListener<T> listener = mReference.get();

        if (listener != null && listener.isAlive()) {
            listener.success(object);
        }
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/networking/ApiCall.kt
package com.garpr.android.networking

import java.lang.ref.WeakReference

class ApiCall<in T>(listener: ApiListener<T>) : ApiListener<T> {

    private val reference = WeakReference<ApiListener<T>>(listener)


    override fun failure(errorCode: Int) {
        val listener = reference.get()

        if (listener != null && listener.isAlive) {
            listener.failure(errorCode)
        }
    }

    override val isAlive: Boolean
        get() = reference.get()?.isAlive == true

    override fun success(`object`: T?) {
        val listener = reference.get()

        if (listener != null && listener.isAlive) {
            listener.success(`object`)
        }
    }

}
